contracts/sol6/Dao/mock/KyberTokenFeeHandler.sol

```diff
@@ -0,0 +1,582 @@
+                .getLatestBRRDataWithCache();
+            require(burnBps.add(rewardBps).add(rebateBps) == BPS, "Bad BRR values");
+            
+            emit BRRUpdated(rewardBps, rebateBps, burnBps, expiryTimestamp, epoch);
+
+            // Update brrAndEpochData
+            updateBRRData(rewardBps, rebateBps, expiryTimestamp, epoch);
+        }
+    }
+
+    function readBRRData()
+        public
+        view
+        returns (
+            uint256 rewardBps,
+            uint256 rebateBps,
+            uint256 expiryTimestamp,
+            uint256 epoch
+        )
+    {
+        rewardBps = uint256(brrAndEpochData.rewardBps);
+        rebateBps = uint256(brrAndEpochData.rebateBps);
+        epoch = uint256(brrAndEpochData.epoch);
+        expiryTimestamp = uint256(brrAndEpochData.expiryTimestamp);
+    }
+
+    function updateBRRData(
+        uint256 reward,
+        uint256 rebate,
+        uint256 expiryTimestamp,
+        uint256 epoch
+    ) internal {
+        // reward and rebate combined values <= BPS. Tested in getBRR.
+        require(expiryTimestamp < 2**64, "expiry timestamp overflow");
+        require(epoch < 2**32, "epoch overflow");
+
+        brrAndEpochData.rewardBps = uint16(reward);
+        brrAndEpochData.rebateBps = uint16(rebate);
+        brrAndEpochData.expiryTimestamp = uint64(expiryTimestamp);
+        brrAndEpochData.epoch = uint32(epoch);
+    }
+
+    function getRRWeiValues(uint256 RRAmountWei)
+        internal
+        returns (
+            uint256 rewardWei,
+            uint256 rebateWei,
+            uint256 epoch
+        )
+    {
+        // Decoding BRR data
+        uint256 rewardInBps;
+        uint256 rebateInBps;
+        (rewardInBps, rebateInBps, epoch) = getBRR();
+
+        rebateWei = RRAmountWei.mul(rebateInBps).div(BPS);
+        rewardWei = RRAmountWei.mul(rewardInBps).div(BPS);
+    }
+
+    function updateRebateValues(
+        uint256 rebateWei,
+        address[] memory rebateWallets,
+        uint256[] memory rebateBpsPerWallet
+    ) internal returns (uint256 totalRebatePaidWei) {
+        uint256 totalRebateBps;
+        uint256 walletRebateWei;
+
+        for (uint256 i = 0; i < rebateWallets.length; i++) {
+            require(rebateWallets[i] != address(0), "rebate wallet address 0");
+
+            walletRebateWei = rebateWei.mul(rebateBpsPerWallet[i]).div(BPS);
+            rebatePerWallet[rebateWallets[i]] = rebatePerWallet[rebateWallets[i]].add(
+                walletRebateWei
+            );
+
+            // a few wei could be left out due to rounding down. so count only paid wei
+            totalRebatePaidWei = totalRebatePaidWei.add(walletRebateWei);
+            totalRebateBps = totalRebateBps.add(rebateBpsPerWallet[i]);
+        }
+
+        require(totalRebateBps <= BPS, "rebates more then 100%");
+    }
+
+    function validateTokenToKncRateToBurn(uint256 rateTokenToKnc) internal view {
+        // No validation for now
```

Comment:
probably need to use chain link with two tokens.
KNC-Eth price feed and other token to Eth

but don't have to do now.

Reply:
yeah, leaving it out for now